rep-2009.rst

```diff
@@ -0,0 +1,446 @@
+Note that there can be ``m`` supported types for each node, where ``m`` is a positive integer.
+For convenience, let's also assume that the supported types are prioritized in their respective order, such that the priority of ``T_1`` is the highest, ``T_2`` is the second highest, and so on.
+
+Using the node notation described above, we can then use the following notation to describe the agreed upon message type between multiple nodes.
+In the example below, node ``1`` (``N_1``) supports only type ``x``, and node ``2`` (``N_2``) supports types ``x`` and ``y``.
+In this case, the selected message type is ``x``, as shown by the ``x`` over the arrow pointing from node ``1`` to node ``2``.
+More practically, ``N_1`` is publishing ``x``, and ``N_2`` is subscribing to a topic with the selected message type ``x``.
+
+.. code-block::
+
+            x
+   N_1(x) ----> N_2(x, y)
+
+
+Simple Examples
+^^^^^^^^^^^^^^^
+
+We can now use this notation to reason about the agreed upon the selected message type in several different scenarios.
+There are several cases that are clear.
+
+.. code-block::
+
+   (1a)
+                    y
+           N_1(y) ----> N_2(x, y)
+
+   (1b)
+                    x
+        N_1(x, y) ----> N_2(x)
+
+   (1c)
+                    y
+        N_1(x, y) ----> N_2(y)
+
+   (1d)
+                    x
+     N_1(x, y, z) ----> N_2(x, a, b)
+
+   (1e)
+                    x
+     N_1(x, y, z) ----> N_2(a, b, x)
+
+
+Publishing to Multiple Nodes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There are also the cases where there are more than two nodes.
+
+In the following case, ``N_3`` has the limiting supported type, ``y``, so ``N_1`` will publish ``y``, despite the fact that both ``N_1`` and ``N_2`` prefer ``x``.
+This is assuming that the function for picking the selected types prioritizes sending one message over sending multiple messages.
+
+.. code-block::
+
+   (2a)
+                 y
+     N_1(x, y) -------> N_2(x, y)
+                    |
+                    |-> N_3(y)
+
+In the following case, the two nodes receiving data from ``N_1`` both require different message types.
+Thus, ``N_1`` has two selected message types, ``x`` and ``y``, and thus ``N_1`` has two publishers.
+
+.. code-block::
+
+   (2b)
+                 x
+     N_1(x, y) ----> N_2(x)
+             |
+             |   y
+             |-----> N_3(y)
+
+
+Negotiating Publisher and Subscriber Pairs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To discuss negotiating publisher and subscriber pairs, we'll have to use additional notation.
+
+Let ``N_p([x, y, z], {x: [x, y, z], y: [y, z, x], z: [z, x, y]})`` be a node ``p`` thats negotiating publishers and subscribers support ``x``, ``y``, and ``z`` message types, as given by the first argument.
+Then let the second argument be a dictionary that maps each message type to an ordered preference.
+This dictionary is used to determine the preference of node ``p``'s subscriber.
```

Comment:
As we discussed offline, I think the explanation of this notation could be a bit clearer.

My understanding of this now is that in a `Negotiating Pair`, the publisher side of the pair will first determine it's type based on the subscribers attached to it.  The types that the publisher can provide are the first list in the notation.  Once that type has been established, then the subscriber side of the pair will provide a list of preferences to it's upstream publisher.  So if the publisher side of the pair negotiated for x, then the subscriber side of the pair will provide the list from the dictionary with the key `x` (the list `[x,y,z]`).  If the publisher side of the pair negotiated for y, then the subscriber side of the pair will provide the list from the dictionary with the key `y` (the list `[y,z,x]`).

I'm not entirely sure how to make that clearer above, but it probably warrants a bit more explanation and maybe more of an example.